src/modules/rlm_imap/rlm_imap.c

```diff
@@ -0,0 +1,301 @@
+	
+	if (randle->result != CURLE_OK) {
+		talloc_free(randle);
+		return RLM_MODULE_REJECT;
+	}
+	
+	talloc_free(randle);
+	return RLM_MODULE_OK;
+}
+
+
+
+
+static rlm_rcode_t CC_HINT(nonnull) mod_authenticate(UNUSED void *instance, void *thread, REQUEST *request)
+{
+	VALUE_PAIR const 		*username;
+	VALUE_PAIR const 		*password;
+	
+	rlm_imap_thread_t       	*t 		= thread;
+	fr_curl_io_request_t    	*randle;
+    
+	fr_time_delta_t        		timeout;
+	CURLcode        		ret;
+
+	randle = fr_curl_io_request_alloc(request);
+	if (!randle){
+		error:
+		return RLM_MODULE_FAIL;
+	}
+
+	/*
+	 *Checks that there is a username and password, and that the password is not blank
+	 */
+	username = fr_pair_find_by_da(request->packet->vps, attr_user_name, TAG_ANY);
+	password = fr_pair_find_by_da(request->packet->vps, attr_user_password, TAG_ANY);
+	if (!username) {
+		REDEBUG("Attribute \"User-Name\" is required for authentication");
+		return RLM_MODULE_INVALID;
+	}
+	if (!password) {
+		RDEBUG2("Attribute \"User-Password\" is required for authentication");
+		return RLM_MODULE_INVALID;
+	}
+	if (password->vp_length == 0) {
+		RDEBUG2("\"User-Password\" must not be empty");
+		return RLM_MODULE_INVALID;
+	}
+	
+	/*
+	 *Set the options for CURL, que the call, and yield until their is a response or a timesout
+	 */
+	SET_OPTION(CURLOPT_USERNAME, username->vp_strvalue);
+	SET_OPTION(CURLOPT_PASSWORD, password->vp_strvalue);
+	SET_OPTION(CURLOPT_URL, imapURI);
+    
+	timeout = 5000;
+	SET_OPTION(CURLOPT_CONNECTTIMEOUT_MS, timeout);
+	SET_OPTION(CURLOPT_TIMEOUT_MS, timeout);
+    
+	ret = fr_curl_io_request_enqueue(t->mhandle, request, randle);
+    
+	return unlang_module_yield(request, mod_authenticate_resume, NULL, randle);
+}
+
+
+static int mod_bootstrap(void *instance, CONF_SECTION *conf)
+{
+	char const        	*name;
+	rlm_imap_t       	*inst = instance;
+	name = cf_section_name2(conf);
+	if (!name) name = cf_section_name1(conf);
+	inst->name = name;
+	if (fr_dict_enum_add_name_next(fr_dict_attr_unconst(attr_auth_type), inst->name) < 0) {
+		PERROR("Failed adding %s alias", attr_auth_type->name);
+		return -1;
+	}
+	inst->auth_type = fr_dict_enum_by_name(attr_auth_type, inst->name, -1);
+	rad_assert(inst->auth_type);
+	return 0;
+}
+
+
+
+static int mod_load(void)
+{
+	if (fr_curl_init() < 0) return -1;
+
+#ifdef HAVE_JSON
+	fr_json_version_print();
+#endif
+	return 0;
+
+}
+static void mod_unload(void)
+{
+	fr_curl_free();
+}
+
+
+static int mod_thread_instantiate(CONF_SECTION const *conf, void *instance, fr_event_list_t *el, void *thread)
+{
+	rlm_imap_thread_t    		*t 	= thread;
+	fr_curl_handle_t    		*mhandle;
+	CONF_SECTION        		*my_conf;
+
+	t->inst = instance;
+
+	my_conf = cf_section_dup(NULL, NULL, conf, cf_section_name1(conf), cf_section_name2(conf), true);
+	talloc_free(my_conf);
+   
+	mhandle = fr_curl_io_init(t, el, false);
+	if (!mhandle) return -1;
+	t->mhandle = mhandle;
+	    
+	fr_curl_free();
```

Comment:
Why are you de-initialising the curl library in the thread instantiate function?